### ISSUE TITLE
Configure Elasticsearch _id dynamically from document

### ISSUE THREAD

**Issue by jfschaff** [django-es/django-elasticsearch-dsl#271]
Hi,

I'm using `django-elasticsearch-dsl` to index two different Django models into a single ES index.

To get a common unique id for both models I use a custom `id` field specially generated for my ES index (I could not rely on individual model ids, as I may have two objects with the same id, or the ids could even be of different types).

The code looks like:
```python
class ModelA(Model):
    [...]


class ModelB(Model):
    [...]


@registry.register_document
@common_index.document
class AToCommonDocument(Document):
    class Django:
        model = ModelA
    id = fields.KeywordField()
    def prepare_id(self, instance):
        return <some_common_id_for_A_and_B>
    [...]
 

@registry.register_document
@common_index.document
class BToCommonDocument(Document):
    class Django:
        model = ModelB
    id = fields.KeywordField()
    def prepare_id(self, instance):
        return <some_common_id_for_A_and_B>
    [...]
```
This works well, and allows me to have a single index and search it efficiently, even if the underlying objects actually come from two tables. The ES entries look like:
```
{
    "_index" : "common",
    "_type" : "_doc",
    "_id" : "2",
    "_score" : 1.0,
    "_source" : {
        "id" : 110,
        [other fields]
    }
}
```
Now, my issue is that I sometimes get two documents with the same ES id (`_id`), so some documents are lost, because they are overwritten when a second, with identical `_id`, is written. I also noticed that the ES `_id` does not have the same value as the `id` field as defined in the documents above.

So the questions:
1. how is the ES `_id` determined if it is neither the Django id, not the id defined in the document? How does an update works, so that the right document is overwritten?
2. is there a way to control the ES `_id` field? For instance, setting `id` and `_id` to be the same?

**Comment by safwanrahman** [comment]
As per the source [code](https://github.com/sabricot/django-elasticsearch-dsl/blob/master/django_elasticsearch_dsl/documents.py#L159-L167), the package is generating the elasticsearch `_id` from the object `pk`.

You can overwrite the method `_prepare_action` in your document and change the `_id` as per your need.
Keep in mind that as this is a internal method and not documented, it can be changed in the future version of this package. Better to write proper test for that.

**Comment by jfschaff** [comment]
@safwanrahman  do you think it would make sens to add a feature for finer control of the ES `_id`, and which would default to the current behavior (object's `pk`)?

I could contribute but we would first need to agree on an API (`_id` field in the document? Special method `determine_id(self, object)`, or something in the `Django` class maybe?).

**Comment by alexgarel** [comment]
I vote +1 on this: introduce a method to compute id, so that it's easier to overload it.

**Comment by safwanrahman** [comment]
@jfschaff Great. I think having an API would be awesome. I am also not so comfortable to suggest people using private API.

**Comment by safwanrahman** [comment]
This should be fixed by #272
